drivers/stepper/gpio_stepper/gpio_stepper_controller.c

```diff
@@ -0,0 +1,218 @@
+
+	K_SPINLOCK(&data->lock) {
+		data->target_position = position;
+		k_work_reschedule(&data->stepper_dwork, K_NO_WAIT);
+	}
+	return 0;
+}
+
+static int gpio_stepper_is_moving(const struct device *dev, bool *is_moving)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+
+	*is_moving = k_work_delayable_is_pending(&data->stepper_dwork);
+	LOG_DBG("Is Motor Moving %d", *is_moving);
+	return 0;
+}
+
+static int gpio_stepper_set_max_velocity(const struct device *dev, uint32_t velocity)
+{
+	if (velocity == 0) {
+		LOG_ERR("Velocity cannot be zero");
+		return -EINVAL;
+	}
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+
+	K_SPINLOCK(&data->lock) {
+		data->delay_in_us = USEC_PER_SEC / velocity;
+	}
+	LOG_DBG("Setting Motor Speed to %d", velocity);
+	return 0;
+}
+
+static int gpio_stepper_get_micro_step_res(const struct device *dev,
+					   enum micro_step_resolution *micro_step_res)
+{
+	ARG_UNUSED(dev);
+	*micro_step_res = STEPPER_FULL_STEP;
+	return 0;
+}
+
+static int gpio_stepper_motor_controller_init(const struct device *dev)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+
+	const struct gpio_stepper_motor_controller_config *config = dev->config;
+
+	data->dev = dev;
+
+	LOG_DBG("Initializing %s gpio_stepper_motor_controller", dev->name);
+	for (uint8_t n_pin = 0; n_pin < NUMBER_OF_GPIO_PINS; n_pin++) {
+		gpio_pin_configure_dt(&config->control_pins[n_pin], GPIO_OUTPUT_INACTIVE);
+	}
+	k_work_init_delayable(&data->stepper_dwork, stepper_work_step_handler);
+	return 0;
+}
+
+static const struct stepper_api gpio_stepper_api = {
+	.move = gpio_stepper_move,
+	.is_moving = gpio_stepper_is_moving,
+	.set_actual_position = gpio_stepper_set_actual_position,
+	.get_actual_position = gpio_stepper_get_actual_position,
+	.set_target_position = gpio_stepper_set_target_position,
+	.set_max_velocity = gpio_stepper_set_max_velocity,
+	.get_micro_step_res = gpio_stepper_get_micro_step_res};
+
+#define GPIO_STEPPER_MOTOR_CONTROLLER_DEFINE(inst)                                                 \
+	static const struct gpio_dt_spec gpio_stepper_motor_control_pins_##inst[] = {              \
+		DT_INST_FOREACH_PROP_ELEM_SEP(inst, gpios, GPIO_DT_SPEC_GET_BY_IDX, (,)),         \
+	};                                                                                         \
+	BUILD_ASSERT(ARRAY_SIZE(gpio_stepper_motor_control_pins_##inst) == NUMBER_OF_GPIO_PINS,    \
+		     "This driver only supports 4 control pins");                                  \
+	static struct gpio_stepper_motor_controller_data                                           \
+		gpio_stepper_motor_controller_data_##inst = {0};               \
+	static const struct gpio_stepper_motor_controller_config                                   \
+		gpio_stepper_motor_controller_config_##inst = {                                    \
+			.control_pins = gpio_stepper_motor_control_pins_##inst};                   \
```

Comment:
```suggestion
			.control_pins = gpio_stepper_motor_control_pins_##inst,     \
			};                   \
```

(with the tab aligned, you got the idea)

Reply:
`clang-format`

Reply:
/me hides